CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/MultiTaskInteraction.java

```diff
@@ -0,0 +1,139 @@
+ * {@link SharedInteractionRunnable} and the MultiTaskInteraction will request updates on the 
+ * completeness of their execution. It is then the MultiTaskInteraction that collates the status of all
+ * running tasks and updates the {@link PluginInteraction} of execution completeness.</p>
+ * 
+ * <p>As {@link PluginInteraction} instances expect to be updated chronologically. Each step should be finished 
+ * before commencing to the next step. To achieve this, the method {@code waitForTasksToComplete()} should 
+ * be called after all tasks are scheduled so that the calling thread is forced to wait for all tasks to complete 
+ * and prevent external calls to the {@link PluginInteraction}.</p>
+ * 
+ * @author capricornunicorn123
+ */
+public class MultiTaskInteraction {
+    private final PluginInteraction interaction;
+    private final List<SharedInteractionRunnable> tasks = new ArrayList<>();
+    
+    
+    public MultiTaskInteraction(final PluginInteraction interaction, final List<SharedInteractionRunnable> tasks){
+        this.interaction = interaction;
+        this.tasks.addAll(tasks);
+    }
+    
+    public MultiTaskInteraction(final PluginInteraction interaction){
+        this.interaction = interaction;
+    }
+    
+    /**
+     * Registers a tasks to this MultiTaskInteraction.
+     * @param task 
+     */
+    public void addTask(final SharedInteractionRunnable task){
+        tasks.add(task);
+    }
+    
+    /**
+     * Determines how many tasks have been registered to this MultiTaskInteraction.
+     * 
+     * @return 
+     */
+    public int getTaskCount(){
+        return tasks.size();
+    }
+    
+    /**
+     * Updates the PluginInteraction of the progress of the running tasks and continues to do so until all tasks are complete.
+     * This method is intended to block further execution on the calling thread until these tasks have completed.
+     * 
+     * @throws InterruptedException 
+     */
+    public void waitForTasksToComplete() throws InterruptedException {
+        // Wait until all tasks are complete
+        while (!isComplete()){
+            setProgress(true);
+        }
+    }
+    
+    /**
+     * Updates the progress of an interaction based on the size and level of completion of registered tasks.
+     * @param cancellable
+     * @throws InterruptedException 
+     */
+    private void setProgress(final boolean cancellable) throws InterruptedException {
+
+        int currentStep = 0;
+        int totalSteps = 0;
+        for (SharedInteractionRunnable task : tasks) {
+            currentStep += task.getCurrentStep();
+            totalSteps += task.getTotalSteps();
+        }
+        interaction.setProgress(currentStep, totalSteps, cancellable);
+        
+        // Give progress reporting mechanisms time to update before the calling class is able to update progress again
+        Thread.sleep(500);
```

Comment:
The Thread.sleep() here is to prevent the waitForTasksToComplete() from being called too frequently. Upon reconsideration, the Thread.sleep() should probably be moved to the waitForTasksToComplete() method as it is the method controlling the waiting. However the Thread.sleep() is needed as excluding it causes recursive updates of the PluginInteraction to happen too frequently resulting in a progress bar that flickers and stalls instead of smoothly loading from left to right. An alternate approach not involving Thread.sleep() would be a while loop and a for loop that arbitrarily counts until threshold and then triggers the updates roughly the same frequency as every 500 ms. If you have any alternate or favored approaches, let me know. Fornow ill move the Thread.sleep() to the waitForTasksToComplete() method.